File: mRemoteV1/Config/Import/ActiveDirectory.cs
using System;
using System.Windows.Forms;
using System.DirectoryServices;
using mRemoteNG.App;
using System.Text.RegularExpressions;
using mRemoteNG.Connection;
using mRemoteNG.Container;
using mRemoteNG.Tree;

namespace mRemoteNG.Config.Import
{
	public class ActiveDirectory
	{
		public static void Import(string ldapPath, TreeNode parentTreeNode)
		{
			try
			{
				var treeNode = ConnectionTreeNode.AddNode(TreeNodeType.Container);
					
				var containerInfo = new ContainerInfo();
				containerInfo.TreeNode = treeNode;
				containerInfo.ConnectionInfo = new ConnectionInfo(containerInfo);
				
				var name = "";
				var match = Regex.Match(ldapPath, "ou=([^,]*)", RegexOptions.IgnoreCase);
				if (match.Success)
				{
					name = match.Groups[1].Captures[0].Value;
				}
				else
				{
					name = Language.strActiveDirectory;
				}
					
				containerInfo.Name = name;
					
				// We can only inherit from a container node, not the root node or connection nodes
				if (ConnectionTreeNode.GetNodeType(parentTreeNode) == TreeNodeType.Container)
				{
					containerInfo.Parent = (ContainerInfo)parentTreeNode.Tag;
				}
				else
				{
					containerInfo.ConnectionInfo.Inheritance.DisableInheritance();
				}
					
				treeNode.Text = name;
				treeNode.Name = name;
				treeNode.Tag = containerInfo;
                Runtime.ContainerList.Add(containerInfo);
					
				ImportComputers(ldapPath, treeNode);
					
				parentTreeNode.Nodes.Add(treeNode);
			}
			catch (Exception ex)
			{
				Runtime.MessageCollector.AddExceptionMessage("Config.Import.ActiveDirectory.Import() failed.", ex, logOnly: true);
			}
		}
			
		private static void ImportComputers(string ldapPath, TreeNode parentTreeNode)
		{
			try
			{
			    const string ldapFilter = "(objectClass=computer)";
					
				var ldapSearcher = new DirectorySearcher();
				var ldapResults = default(SearchResultCollection);
				var ldapResult = default(SearchResult);
					
				ldapSearcher.SearchRoot = new DirectoryEntry(ldapPath);
				ldapSearcher.PropertiesToLoad.AddRange(new[] {"securityEquals", "cn"});
				ldapSearcher.Filter = ldapFilter;
				ldapSearcher.SearchScope = SearchScope.OneLevel;
					
				ldapResults = ldapSearcher.FindAll();
					
				foreach (SearchResult tempLoopVar_ldapResult in ldapResults)
				{
					ldapResult = tempLoopVar_ldapResult;
					var with_2 = ldapResult.GetDirectoryEntry();
					var displayName = Convert.ToString(with_2.Properties["cn"].Value);
					var description = Convert.ToString(with_2.Properties["Description"].Value);
					var hostName = Convert.ToString(with_2.Properties["dNSHostName"].Value);
						
					var treeNode = ConnectionTreeNode.AddNode(TreeNodeType.Connection, displayName);
						
					var connectionInfo = new ConnectionInfo();
					var inheritanceInfo = new ConnectionInfoInheritance(connectionInfo, true);
					inheritanceInfo.Description = false;
					if (parentTreeNode.Tag is ContainerInfo)
					{
						connectionInfo.Parent = (ContainerInfo)parentTreeNode.Tag;
					}
					connectionInfo.Inheritance = inheritanceInfo;
					connectionInfo.Name = displayName;
					connectionInfo.Hostname = hostName;
					connectionInfo.Description = description;
					connectionInfo.TreeNode = treeNode;
					treeNode.Name = displayName;
					treeNode.Tag = connectionInfo; //set the nodes tag to the conI
					//add connection to connections
                    Runtime.ConnectionList.Add(connectionInfo);
						
					parentTreeNode.Nodes.Add(treeNode);
				}
			}
			catch (Exception ex)
			{
				Runtime.MessageCollector.AddExceptionMessage("Config.Import.ActiveDirectory.ImportComputers() failed.", ex, logOnly: true);
			}
		}
	}
}

File: mRemoteV1/Config/Import/ActiveDirectory.vb
Imports System.DirectoryServices
Imports System.Text.RegularExpressions
Imports mRemote3G.App
Imports mRemote3G.Container
Imports mRemote3G.Tree

Namespace Config.Import
    Public Class ActiveDirectory
        Public Shared Sub Import(ldapPath As String, parentTreeNode As TreeNode)
            Try
                Dim treeNode As TreeNode = Node.AddNode(Node.Type.Container)

                Dim containerInfo As New Info()
                containerInfo.TreeNode = treeNode
                containerInfo.ConnectionInfo = New Connection.Info(containerInfo)

                Dim name As String
                Dim match As Match = Regex.Match(ldapPath, "ou=([^,]*)", RegexOptions.IgnoreCase)
                If match.Success Then
                    name = match.Groups(1).Captures(0).Value
                Else
                    name = Language.Language.strActiveDirectory
                End If

                containerInfo.Name = name

                ' We can only inherit from a container node, not the root node or connection nodes
                If Node.GetNodeType(parentTreeNode) = Node.Type.Container Then
                    containerInfo.Parent = parentTreeNode.Tag
                Else
                    containerInfo.ConnectionInfo.Inherit.TurnOffInheritanceCompletely()
                End If

                treeNode.Text = name
                treeNode.Name = name
                treeNode.Tag = containerInfo
                Runtime.ContainerList.Add(containerInfo)

                ImportComputers(ldapPath, treeNode)

                parentTreeNode.Nodes.Add(treeNode)
            Catch ex As Exception
                Runtime.MessageCollector.AddExceptionMessage("Config.Import.ActiveDirectory.Import() failed.", ex, ,
                                                             True)
            End Try
        End Sub

        Private Shared Sub ImportComputers(ldapPath As String, parentTreeNode As TreeNode)
            Try
                Dim strDisplayName, strDescription, strHostName As String

                Const ldapFilter = "(objectClass=computer)"

                Dim ldapSearcher As New DirectorySearcher
                Dim ldapResults As SearchResultCollection
                Dim ldapResult As SearchResult

                With ldapSearcher
                    .SearchRoot = New DirectoryEntry(ldapPath)
                    .PropertiesToLoad.AddRange({"securityEquals", "cn"})
                    .Filter = ldapFilter
                    .SearchScope = SearchScope.OneLevel
                End With

                ldapResults = ldapSearcher.FindAll()

                For Each ldapResult In ldapResults
                    With ldapResult.GetDirectoryEntry()
                        strDisplayName = .Properties("cn").Value
                        strDescription = .Properties("Description").Value
                        strHostName = .Properties("dNSHostName").Value
                    End With

                    Dim treeNode As TreeNode = Node.AddNode(Node.Type.Connection, strDisplayName)

                    Dim connectionInfo As New Connection.Info()
                    Dim inheritanceInfo As New Connection.Info.Inheritance(connectionInfo, True)
                    inheritanceInfo.Description = False
                    If TypeOf parentTreeNode.Tag Is Info Then
                        connectionInfo.Parent = parentTreeNode.Tag
                    End If
                    connectionInfo.Inherit = inheritanceInfo
                    connectionInfo.Name = strDisplayName
                    connectionInfo.Hostname = strHostName
                    connectionInfo.Description = strDescription
                    connectionInfo.TreeNode = treeNode
                    treeNode.Name = strDisplayName
                    treeNode.Tag = connectionInfo 'set the nodes tag to the conI
                    'add connection to connections
                    Runtime.ConnectionList.Add(connectionInfo)

                    parentTreeNode.Nodes.Add(treeNode)
                Next
            Catch ex As Exception
                Runtime.MessageCollector.AddExceptionMessage("Config.Import.ActiveDirectory.ImportComputers() failed.",
                                                             ex, , True)
            End Try
        End Sub

        Private Sub New()
        End Sub
    End Class
End Namespace